### PR TITLE
Refactors describe command to split flags from options

### PR DESCRIPTION
**Pull request by PuneetPunamiya**
Refactors Describe command from umbrella issue --> https://github.com/kubernetes/kubectl/issues/1046

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Refactors describe command to split flags from options

#### Special notes for your reviewer:
```NONE```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go [hunk 1]
```diff
@@ -69,6 +69,29 @@ var (
...
+}
+
+// NewDescribeFlags returns a default DescribeFlags
+func NewDescribeFlags(f cmdutil.Factory, streams genericclioptions.IOStreams) *DescribeFlags {
+	return &DescribeFlags{
+		Factory:           f,
+		Selector:          "",
+		AllNamespaces:     false,
+		FilenameOptions:   &resource.FilenameOptions{},
+		DescriberSettings: &describe.DescriberSettings{},
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:90]: You are changing the functionality in here comparing to https://github.com/kubernetes/kubernetes/pull/112714/files#diff-22297d1a66ecc1d0164cf7d38d7e218aa9163d5f507233e631c9e1f0c65d258cL95

File: staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go [hunk 2]
```diff
@@ -88,19 +111,57 @@ type DescribeOptions struct {
...
+	cmdutil.AddFilenameOptionFlags(cmd, flags.FilenameOptions, "containing the resource to describe")
+	cmdutil.AddLabelSelectorFlagVar(cmd, &flags.Selector)
+	cmd.Flags().BoolVarP(&flags.AllNamespaces, "all-namespaces", "A", flags.AllNamespaces, "If present, list the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace.")
+	cmd.Flags().BoolVar(&flags.DescriberSettings.ShowEvents, "show-events", flags.DescriberSettings.ShowEvents, "If true, display events related to the described object.")
+	cmdutil.AddChunkSizeFlag(cmd, &flags.DescriberSettings.ChunkSize)
+}
 
-		CmdParent: parent,
+// ToOptions converts from CLI inputs to runtime input
+func (flags *DescribeFlags) ToOptions(parent string, args []string) (*DescribeOptions, error) {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:124]: Where is builder initialization? `o.NewBuilder = f.NewBuilder`

File: staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go [hunk 3]
```diff
@@ -109,48 +170,15 @@ func NewCmdDescribe(parent string, f cmdutil.Factory, streams genericclioptions.
 		Example:               describeExample,
 		ValidArgsFunction:     completion.ResourceTypeAndNameCompletionFunc(f),
 		Run: func(cmd *cobra.Command, args []string) {
-			cmdutil.CheckErr(o.Complete(f, cmd, args))
-			cmdutil.CheckErr(o.Validate())
+			o, err := flags.ToOptions(parent, args)
+			cmdutil.CheckErr(err)
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:174]: I'd prefer to preserve `Validate` function even if it does not validate anything.

**Review by ardaguclu — COMMENTED** [review]
Thanks for investing your time to this PR!

File: staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go (hunk 2 above)
Comment (PuneetPunamiya) [reply staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:124]: So I'm directly assigning it here https://github.com/kubernetes/kubernetes/pull/112714/files#diff-22297d1a66ecc1d0164cf7d38d7e218aa9163d5f507233e631c9e1f0c65d258cR150, should I add it as a variable

File: staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go (hunk 3 above)
Comment (PuneetPunamiya) [reply staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:174]: Yeah sounds good 👍🏻

Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:177]: Where is this `cmdutil.CheckErr(o.Validate())` code block?

**Review by ardaguclu — COMMENTED** [review]
Looks good to me, just a minor comment

**Comment by ardaguclu** [comment]
You need to change "fixes" statement in description. Otherwise, when this PR is merged, issue will be closed automatically.

File: staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go (hunk 3 above)
Comment (PuneetPunamiya) [reply staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:177]: I have updated the patch for that, thanks!

**Comment by PuneetPunamiya** [comment]
@ardaguclu I have addressed the comments and updated the patch, thanks for the quick review 🤗

File: staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go [hunk 4]
```diff
@@ -69,6 +69,32 @@ var (
...
+	FilenameOptions   *resource.FilenameOptions
+	DescriberSettings *describe.DescriberSettings
+	genericclioptions.IOStreams
+}
+
+// NewDescribeFlags returns a default DescribeFlags
+func NewDescribeFlags(f cmdutil.Factory, streams genericclioptions.IOStreams) *DescribeFlags {
+	return &DescribeFlags{
+		Factory:         f,
+		Selector:        "",
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:87]: Maybe we can remove these `Selector` and `AllNamespaces` fields setting, their default is already empty and false?

Comment (PuneetPunamiya) [reply staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:87]: yeah make sense 👍🏻

**Comment by MadhavJivrajani** [comment]
/assign @soltysh @eddiezane 
for approval

**Review by seans3 — COMMENTED** [review]
Ready to approve. One question on a small nit pick. This looks good in separating the `Options` (`DescribeOptions`) from the `Cobra` command-line stuff, allowing easier re-use of the describe code. One small nit is that I would have moved the `DescribeOptions` struct to the bottom of the file so `DescribeOptions` and its methods are grouped together. If you don't think that is necessary or helpful, please let me know and I'll approve without that change. Thanks for investing in this cleanup.

**Comment by PuneetPunamiya** [comment]
> Ready to approve. One question on a small nit pick. This looks good in separating the `Options` (`DescribeOptions`) from the `Cobra` command-line stuff, allowing easier re-use of the describe code. One small nit is that I would have moved the `DescribeOptions` struct to the bottom of the file so `DescribeOptions` and its methods are grouped together. If you don't think that is necessary or helpful, please let me know and I'll approve without that change. Thanks for investing in this cleanup.

yeah, I think that make sense, I'll update the patch, thanks for the review 🙂

**Review by seans3 — APPROVED** [review]
/lgtm
/approve

File: staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go [hunk 5]
```diff
@@ -69,38 +69,81 @@ var (
...
-	BuilderArgs []string
-
-	EnforceNamespace bool
-	AllNamespaces    bool
-
-	DescriberSettings *describe.DescriberSettings
+// DescribeFlags directly reflect the information that CLI is gathering via flags. They will be converted to Options,
+// which reflect the runtime requirements for the command.
+type DescribeFlags struct {
+	Factory           cmdutil.Factory
```
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:75]: This is wrong, `DescribeFlags` should ONLY contain flags exposed by the command, `Factory`'s place is only at `DescribeOptions` struct.

Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:75]: Also you should be easily replace the `Factory` with `ResourceBuilderFlags` and `RESTClientGetter`, see how https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go is structured, that's what I'd want to see applied across the board.

File: staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go [hunk 6]
```diff
@@ -69,38 +69,81 @@ var (
...
+	if err != nil {
+		return nil, err
 	}
 
+	if flags.AllNamespaces {
+		enforceNamespace = false
+	}
+
+	if len(args) == 0 && cmdutil.IsFilenameSliceEmpty(flags.FilenameOptions.Filenames, flags.FilenameOptions.Kustomize) {
+		return nil, fmt.Errorf("You must specify the type of resource to describe. %s\n", cmdutil.SuggestAPIResources(parent))
```
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:119]: Errors should be always start with a lower case.

**Review by soltysh — COMMENTED** [review]
Apologies for being late to this, but please open a followup with the fixes and ping me on slack to get it reviewed. Also make sure to apply these changes (where they apply) to your other PRs modifying these bits of code. Thank you!

File: staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go (hunk 5 above)
Comment (PuneetPunamiya) [reply staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:75]: Yeah thanks for addressing this, I'll create a follow up pr on this, but I guess we'll also have to refactor some commands for which changes are already done and merged. For example :- [apply command](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go#L55) also uses `Factory`

Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:75]: Thanks @PuneetPunamiya for noticing that

@soltysh I removed the Factory in the struct in this PR https://github.com/kubernetes/kubernetes/pull/114030

File: staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go [hunk 7]
```diff
@@ -69,38 +69,81 @@ var (
...
+		Selector:          flags.Selector,
+		Namespace:         namespace,
+		Describer:         describer,
+		NewBuilder:        flags.Factory.NewBuilder,
+		BuilderArgs:       builderArgs,
+		EnforceNamespace:  enforceNamespace,
+		AllNamespaces:     flags.AllNamespaces,
+		FilenameOptions:   flags.FilenameOptions,
+		DescriberSettings: flags.DescriberSettings,
+		IOStreams:         flags.IOStreams,
```
Comment (RainbowMango) [inline staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go:138]: Did here forget to set `DescribeOptions.CmdParent` here?

**Pull request merged**